### ISSUE TITLE
`BikeshedIntrinsicFrom` permits lifetime extension

### ISSUE THREAD

**Issue by jswrenn** [rust-lang/rust#129097]
I tried this code:

```rust
#![feature(transmutability, core_intrinsics)]

use std::mem::{Assume, BikeshedIntrinsicFrom};

fn extend<'a>(src: &'a u8) -> &'static u8 {
    transmute::<&'a u8, &'static u8>(src)
}

fn transmute<Src, Dst>(src: Src) -> Dst
where
    Dst: BikeshedIntrinsicFrom<Src>,
{
    unsafe { core::intrinsics::transmute_unchecked(src) }
}


fn main() {}
```

I expected to see this to fail with a compilation error. Instead it succeeded.

### Meta

`rustc --version --verbose`:
```
rustc 1.82.0-nightly (3e9bd8b56 2024-08-08)
binary: rustc
commit-hash: 3e9bd8b566a47c5d1c1dbc7e043b4b7fa5330eca
commit-date: 2024-08-08
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0
```

**Comment by jswrenn** [comment]
It looks like this was just never implemented when support for references was added, rather than this being a case of a subtle bug. I've begun to address this on [a branch](https://github.com/jswrenn/rust/tree/transmute-lifetimes) by modifying `confirm_transmutability_candidate` to emit outlives obligations: https://github.com/jswrenn/rust/blob/6176aafb7a9568ac1a78f16b4f9fba703df4cd55/compiler/rustc_trait_selection/src/traits/select/confirmation.rs#L290-L455

I'm emitting outlives obligations in two places. First, where we handle `IfTransmutable` answers: https://github.com/jswrenn/rust/blob/6176aafb7a9568ac1a78f16b4f9fba703df4cd55/compiler/rustc_trait_selection/src/traits/select/confirmation.rs#L397-L403

This should be sufficient. The other kinds of obligations emitted by the `reference_obligations` helper *are* exercised, and errors arise from violating them.

For some reason, it's not. I've also added these obligations at a top level in `confirm_transmutability_candidate`, peaking into the `src` and `dst` types and eagerly emitting reference obligations if needed: https://github.com/jswrenn/rust/blob/6176aafb7a9568ac1a78f16b4f9fba703df4cd55/compiler/rustc_trait_selection/src/traits/select/confirmation.rs#L431-L443

This is sufficient for catching 'top level' lifetime extensions; e.g.:
```rust
fn extend<'a>(src: &'a u8) -> &'static u8 {
    transmute(src) // error, yay!
}

fn transmute<Src, Dst>(src: Src) -> Dst
where
    Dst: BikeshedIntrinsicFrom<Src>,
{
    unsafe { core::intrinsics::transmute_unchecked(src) }
}
```
...but not nested lifetime extensions, like:
```rust


fn extend_nested<'a>(src: (&'a u8, u8)) -> (&'static u8, u8) {
    transmute(src) // no error :(
}
```
